### ISSUE TITLE
Memory leak - Polars with dash/flask

### ISSUE THREAD

**Issue by PerovskiteCell** [pola-rs/polars#18074]
### Checks

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python

import pathlib, os, shutil, psutil
import polars as pl, pandas as pd, numpy as np

from dash import Dash, dcc, html, Input, State, Output, Patch
import plotly.graph_objects as go

polars_check = True ### Whether the example operates with polars or pandas.

#Constants
interval_time = 5E2 # (ms)
process = psutil.Process(os.getpid())

folder = pathlib.Path(r'C:\PerovskiteCell example')

n_files = 100 #Number of files in folder
n_lines = 500000 #Number of total lines in folder
n_cols = 25


#Generating sample data in example folder (Only once).
if not folder.exists():

    size = int(n_lines / n_files)
    col = np.linspace(-1E3, 1E3, num=size)

    df = pl.DataFrame({f'col{n}': col for n in range(n_cols)})

    # Creating folder & files
    os.makedirs(folder)

    f_path0 = folder.joinpath('0.parquet')
    df.write_parquet(f_path0)

    for n in range(1, n_files):
        shutil.copy2(f_path0, folder.joinpath(f'{n}.parquet'))


#Functions
def pl_data():
    """Retrieves data via the polars route"""
    
    lf = pl.scan_parquet([folder.joinpath(f'{n}.parquet') for n in range(n_files)])
    lf = lf.select(pl.col('col0', 'col1').gather([n * int(n_lines / n_files + 1) for n in range(n_files)]))

    return lf.collect()


def pd_data():
    """Retrieves data via the pandas route"""

    df = (pd.read_parquet(folder.joinpath(f'{n}.parquet'), columns=['col0', 'col1'])
                                         for n in range(n_files))
    
    df = pd.concat(df, ignore_index=True)
    df = df.iloc[[n * int(n_lines / n_files + 1) for n in range(n_files)],]
    
    return df



#App (initialization)
app = Dash()
app.layout = html.Div([dcc.Graph(id='graph', figure=go.Figure({'layout': {'title': 'Polars' if polars_check else 'Pandas',
                                                                          'xaxis': {'title': 'Nr. of callbacks (#)'},
                                                                          'yaxis': {'title': 'Mem usage (MB)'}
                                                                          }})),
                        dcc.Interval(id = 'check', 
                                        interval = interval_time,
                                        max_intervals = 250),

                        html.Button('Save data', id='save_button', n_clicks=0),
                        dcc.Download(id="download")])



@app.callback(
    Output('graph', 'figure'),
    Input('check', 'n_intervals'),

    running=[(Output("save_button", "disabled"), True, False)],
    prevent_initial_callback=True)

def plot(N):

    # Data retrieval
    if polars_check:
        df = pl_data()
    else:
        df = pd_data()

    #Plotting
    fig = Patch()
    fig['data'][0]['x'].append(N)
    fig['data'][0]['y'].append(process.memory_info().rss / (1024 **2))

    return fig


@app.callback(
    Output('download', 'data'),

    Input('save_button', 'n_clicks'),
    State('graph', 'figure'),

    prevent_initial_callback=True)

def save(_, fig):

    data = pl.DataFrame({para: fig['data'][0][para] for para in ['x', 'y']})

    return dict(content=data.write_ndjson(None), filename=f'{"Polars" if polars_check else "Pandas"}.csv')




if __name__ == '__main__':
    app.run(debug=False, port = 8050)

```

### Issue description

![Memory comparison](https://github.com/user-attachments/assets/679dc1d9-effd-4924-b483-82ec1910501e)

Memory-use of the python-process (in Windows 10) keeps gradually increasing with each successive callback, if polars is used in a callback. When pandas is used, the memory-usage stabilizes.

Issue is probably rooted in the combination of flask/polars, as [somebody else just using flask also encountered it](https://stackoverflow.com/questions/78831942/polars-memory-usage-just-building-up-and-not-coming-down/78834890#78834890).

(If somebody knows how to fix this, or work around this, **please let me know.** I love the syntax/performance of polars.)

### Expected behavior

The memory-use should stabilize; like it happens for the callback with pandas.

### Installed versions

<details>

```
--------Version info---------
Polars:               1.4.1
Index type:           UInt32
Platform:             Windows-10-10.0.19045-SP0
Python:               3.11.3 (tags/v3.11.3:f3909b8, Apr  4 2023, 23:49:59) [MSC v.1934 64 bit (AMD64)]

----Optional dependencies----
adbc_driver_manager:  <not installed>
cloudpickle:          <not installed>
connectorx:           <not installed>
deltalake:            <not installed>
fastexcel:            <not installed>
fsspec:               2023.9.2
gevent:               <not installed>
great_tables:         <not installed>
hvplot:               <not installed>
matplotlib:           3.9.0
nest_asyncio:         1.5.6
numpy:                1.25.1
openpyxl:             3.1.2
pandas:               2.2.2
pyarrow:              16.0.0
pydantic:             2.8.2
pyiceberg:            <not installed>
sqlalchemy:           <not installed>
torch:                <not installed>
xlsx2csv:             <not installed>
xlsxwriter:           <not installed>
```

</details>

**Comment by ritchie46** [comment]
Can someone confirm on linux/macos? I cannot reproduce on linux.

**Comment by ritchie46** [comment]
@PerovskiteCell can you try setting `POLARS_MAX_THREADS=1`?

**Comment by PerovskiteCell** [comment]
@ritchie46 
![newplot (17)](https://github.com/user-attachments/assets/af98d9e2-15fc-4191-b79f-79e7b7f97d3b)

`os.environ['POLARS_MAX_THREADS'] = '1'`

I now added the above just below the import statements.
_(Correct me if I'm wrong; I'm quite an inexperienced, part-time programmer.)_

**Comment by itamarst** [comment]
I can't reproduce on Linux either (tried with 1.4.1 and with compiled debug version from a couple of weeks ago).

**Comment by itamarst** [comment]
What happens if you keep running it? Does it keep growing forever, or does it stabilize eventually? On Linux memory usage is around 200MB, and that graph ends at 250MB, so I'm wondering if it might e.g. stabilize just at a higher point.

**Comment by itamarst** [comment]
I tried running with mimalloc on Linux. Could not reproduce either. Doesn't mean it won't potentially have issues with mimalloc on Windows, though.

**Comment by PerovskiteCell** [comment]
@itamarst
![newplot (19)](https://github.com/user-attachments/assets/4f87cdf5-9f3b-4e87-9e7e-dc98e949b087)

It keeps expanding linearly. It doesn't seem to stabilize.

**Comment by ritchie46** [comment]
Do we still use mimalloc on windows? Maybe we should try a release without it.

**Comment by deanm0000** [comment]
I think the workaround would be to run `pl_data()` in a separate process. Something like

```
import multiprocessing
import polars as pl

def pl_data(df):
    df['df']=pl.select(a=1)

@app.callback(...)
def other_func():
    manager = multiprocessing.Manager()
    df=manager.dict()
    p = multiprocessing.Process(target=pl_data, args=(df,))
    p.start()
    p.join()
    df=df['df']
```

Note that in multiprocessing, to get the return value you have to make a shared object that is passed to the function which it adds to. That might be enough to resolve the issue since the main process is still handling a df. If that's the case then I'd set it up so that you don't even import polars in the main, move the import to the function and have the function return the serialization (ie. write_json())

so something like

```
def pl_data(df):
    import polars as pl
    df['df']=pl.select(a=1).write_json()

@app.callback(...)
def other_func():
    manager = multiprocessing.Manager()
    df=manager.dict()
    p = multiprocessing.Process(target=pl_data, args=(df,))
    p.start()
    p.join()
    return df['df'], 200
```

If even that still holds memory then that would be really bizarre I think.

**Comment by itamarst** [comment]
Do you get the same effect when you use CSV and scan_csv, rather than Parquet?

**Comment by PerovskiteCell** [comment]
@itamarst

Yes. I originally[ had the issue with csv](https://community.plotly.com/t/dash-polars-ram-use-keeps-increasing/86041/3).  I modified the reproducible example to parquet to get statistics quicker (parquet processes faster than csv).

**Comment by itamarst** [comment]
Is the original reproducer you copy/pasted exactly the same as what you're running locally? I was able to reproduce memory leaking by combining the two examples (the one above, and the linked CSV one) _if_ I actually graphed data from the dataframe, which the example above does not.

Assuming that graphing is indeed the issue, the code that's leaking is:

```python
trace = go.Scattergl(x=df["col0"], y=df["col1"], mode="lines+markers")
```

Where `go` is a plotly object. So this implies a leak either in plotly, or in the way that Polars data is converted into something plotly can use.

**Comment by deanm0000** [comment]
plotly expects 

> Type: list, numpy array, or Pandas series of numbers, strings, or datetimes.
https://plotly.com/python/reference/scattergl/#scattergl-x

but you're giving it pl.Series. I'd guess it's falling back on duck typing that it is iterable to make it work.

Try doing

```
trace = go.Scattergl(x=df["col0"].to_numpy(), y=df["col1"].to_numpy(), mode="lines+markers")
```

**Comment by itamarst** [comment]
@deanm0000 The suggestion is that there's memory leak even without graphing. So this will need a developer with a Windows machine to reproduce and then figure out.

**Comment by deanm0000** [comment]
@itamarst I was just going off where you said that's where the memory leak is. ¯\_(ツ)_/¯

**Comment by itamarst** [comment]
I'm getting vibes that there's _two_ memory leaks in different places. For the one you're talking about, also should check if they use the dataframe interchange protocol (I think plotly might?).

**Comment by ritchie46** [comment]
Can you try setting this env var before Polars is imported? `MIMALLOC_RESET_DECOMMITS=1`?

**Comment by PerovskiteCell** [comment]
@ritchie46 

I played a bit more around with the environment options of mimalloc.

![newplot (24)](https://github.com/user-attachments/assets/71c32293-7c82-429e-b4db-9939bb3625f8)

Both `MIMALLOC_PURGE_DELAY = 0` and `MIMALLOC_ABANDONED_PAGE_RESET = 1` minimize the memory leak; `MIMALLOC_RESET_DECOMMITS = 1` doesn't give a noticeable difference.

`MIMALLOC_ABANDONED_PAGE_RESET = 1` works the best, since it is the quickest. This works good enough for my project.

@all; thanks for the help & suggestions!

**Comment by ritchie46** [comment]
Right, I then think this is due to fragmentation in the `mimalloc` allocator. Maybe we should default to the standard allocator for windows.

**Comment by ritchie46** [comment]
I believe we don't use MIMALLOC anymore on windows. Could you check if the issue is still present?

**Comment by PerovskiteCell** [comment]
@ritchie46 
![issue_resolved](https://github.com/user-attachments/assets/42ee3451-c89f-49b7-87f7-3466a6b18c67)

Issue seems to be completely resolved now in windows-11 (checked with polars-1.18.0). Nice!

**Comment by ritchie46** [comment]
Great!
